packages/hardhat-core/src/internal/solidity/compiler/downloader.ts

```diff
@@ -159,7 +159,19 @@ export class CompilerDownloader {
       await this.downloadCompilersList(platform);
     }
 
-    return fsExtra.readJson(this.getCompilersListPath(platform));
+    try {
+      return await fsExtra.readJSON(this.getCompilersListPath(platform));
+    } catch (error) {
+      // if parsing throws a syntax error, redownload and parse once more
+      if (!(error instanceof SyntaxError)) {
+        // eslint-disable-next-line @nomiclabs/hardhat-internal-rules/only-hardhat-error
```

Comment:
My intention here was to touch nothing unless it is explicitly the parsing error, so rethrowing ... but that does mean running against the eslint rule.

Reply:
I like this change!

I also think that we should implement the log to download to a tmp file and move once ready, as a general helper, because we download other things that are harder to validate. For example, the compiler's binaries.

Reply:
I have modified `packages/hardhat-core/src/internal/util/download.ts` to download to a temp file in the same directory are the intended final filepath so `compilers/list.json` will initially be downloaded to `compilers/tmp-list.json`; on successful download the file is then moved into the intended final filepath.

I can only find two usages of download - download the compiler list and download an individual compiler. So I have made the tmp file change directly in `download.ts` but I can change that to be a separate function `download-via-tmp` or whatever and apply it on compilers list initially. What is your preference?

Reply:
I think this makes sense